proposals/kapp/001-crd-upgrade-safety/README.md

```diff
@@ -0,0 +1,95 @@
+- Fail Open - In the scope of this document, "failing open" is used to represent that if we can't explicitly validate something then we should succeed.
+- Preflight Check - A preflight check is a validation that is run prior to performing some action. Typically, if the validation fails the action will not be performed.
+
+## Proposal
+In order to address the problem identified, it is proposed that a new, optional, preflight check is added to `kapp` to enable verifying that an `update` operation on a CRD is safe before attempting to apply any changes to the cluster.
+
+### Goals and Non-goals
+
+#### Goals
+- Providing `kapp` users a way to enable verification that an `update` operation on CRDs is safe and won't cause data loss or break clients
+
+#### Non-goals
+- Anything not pertaining to providing `kapp` users a way to ensure that `update` operations on CRDs are safe before applying changes
+
+### Specification
+A new preflight check will be added to `kapp` called `CRDUpgradeSafety`.
+
+#### `preflight` Package Changes
+The `CRDUpgradeSafety` preflight check is a very nuanced preflight check and different users will likely want to be able to configure this check a bit further to do things like:
+- Only issue warnings about potential problems
+- Fail closed/open
+
+In order to accomodate this configurability without introducing various versions of the same preflight check, the logic in https://github.com/carvel-dev/kapp/tree/develop/pkg/kapp/preflight will be updated to allow for configurable preflight checks.
+
+As an example, configuring a preflight check could look something like:
+```sh
+kapp deploy ... --preflight=Check=true:{"mode":"strict"}
+```
+
+To provide a high quality user experience, each configurable preflight check should have a sane default and not require a user to provide any configuration information.
+
+#### `CRDUpgradeSafety` Logic
+Updates to a CRD can be deemed unsafe if any of the following occur:
+- An existing version that is a stored version in the object store is removed
+- Scope of a CRD is changed from Namespace to Cluster (or vice versa)
+- Removing a field in an existing version
+- Adding a new required field
+- Updates to an existing field that makes validation more restrictive or would break existing versions/clients. Some brief examples: changing a field's type from string to int, increasing minimum value, removing an enum value.
+
+The most straightforward changes to check for are the first four and will have explicit checks in the implementation.
+
+Checking for breaking changes to an existing field is much more nuanced and is more difficult to catch all valid cases.
+It is proposed that an iterative approach is taken where the base implementation does not contain any specific checks against changed fields, but is written in a way that it is easy to add checks in the future.
```

Comment:
This sounds like we need a graduation plan for this feature.

If `kapp` makes a release while we are still iterating, can we make the preflight check more restrictive in a future version? Should we put immature preflight check implementations behind a feature gate? 

I don't know if kapp has any mechanism for this kind of thing. Just wondering if there's a user-friendly way to roll this out.

Reply:
`kapp` is still v0 - latest release is v0.60.0 - so following semver it should be okay if a release is cut while still iterating. A feature gate probably isn't necessary because it is an opt-in preflight check. The preflight check will be disabled by default.

Reply:
This comment has been sitting for a little while. Going to mark as resolved. Please feel free to unresolve if you feel it has been insufficiently addressed.